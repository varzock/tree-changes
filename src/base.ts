import * as equal from 'fast-deep-equal';
import {
  compareNumbers,
  compareValues,
  getIterables,
  getDataValues,
  isArray,
  isCompatible,
  isDefined,
  isPlainObject,
  nested,
} from './helpers';

import { Data, Key, Value, TreeChanges } from './types';

export default function treeChanges(previousData: Data, data: Data): TreeChanges {
  if (!previousData || !data) {
    throw new Error('Missing required parameters');
  }

  const added = (key?: Key, value?: Value): boolean => {
    try {
      // skip if immediate parent isn't an array or object
      if (!isCompatible(previousData, data, { key })) {
        return false;
      }

      const [left, right] = getDataValues(previousData, data, { filter: true, key });

      if (value) {
        // skip if key exists in the narrowed data
        if (key && left[key] && right[key]) {
          return false;
        }

        return compareValues(left, right, value);
      }

      if (key && !(key in left) && key in right) {
        return true;
      }

      const [leftIterable, rightIterable] = getIterables(previousData, data, {
        key,
      });

      return leftIterable.length < rightIterable.length;
    } catch (error) {
      return false;
    }
  };

  const changed = (key?: Key, actual?: Value, previous?: Value): boolean => {
    const left = nested(previousData, key);
    const right = nested(data, key);
    const hasActual = isDefined(actual);
    const hasPrevious = isDefined(previous);

    if (hasActual && !hasPrevious) {
      const leftComparator = isArray(actual) ? actual.indexOf(left) < 0 : left !== actual;
      const rightComparator = isArray(actual) ? actual.indexOf(right) >= 0 : right === actual;

      return leftComparator && rightComparator;
    }

    if (hasPrevious) {
      const leftComparator = isArray(previous) ? previous.indexOf(left) >= 0 : left === previous;
      const rightComparator = isArray(actual) ? actual.indexOf(right) >= 0 : right === actual;

      return leftComparator && rightComparator;
    }

    if ([left, right].every(isArray) || [left, right].every(isPlainObject)) {
      return !equal(left, right);
    }

    return left !== right;
  };

  const changedFrom = (key: Key, previous: Value, actual?: Value): boolean => {
    if (!isDefined(key)) {
      return false;
    }

    const left = nested(previousData, key);
    const right = nested(data, key);
    const hasActual = isDefined(actual);
    const leftComparator = isArray(previous) ? previous.indexOf(left) >= 0 : left === previous;
    const rightComparator = isArray(actual) ? actual.indexOf(right) >= 0 : right === actual;

    return leftComparator && (hasActual ? rightComparator : !hasActual);
  };

  /**
   * @deprecated
   */
  const changedTo = (key: Key, actual: Value): boolean => {
    if (!isDefined(key)) {
      return false;
    }

    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'development') {
      // eslint-disable-next-line no-console
      console.warn('`changedTo` is deprecated! Replace it with `change`');
    }

    return changed(key, actual);
  };

  const decreased = (key: Key, actual?: Value, previous?: Value): boolean => {
    if (!isDefined(key)) {
      return false;
    }

    return compareNumbers(previousData, data, 'decreased', { key, actual, previous });
  };

  const emptied = (key?: Key): boolean => {
    try {
      const [left, right] = getIterables(previousData, data, { includeStrings: true, key });
      return !!left.length && !right.length;
    } catch (error) {
      return false;
    }
  };

  const filled = (key?: Key): boolean => {
    try {
      const [left, right] = getIterables(previousData, data, {
        filter: false,
        includeStrings: true,
        key,
      });

      return !left.length && !!right.length;
    } catch (error) {
      return false;
    }
  };

  const increased = (key: Key, actual?: Value, previous?: Value): boolean => {
    if (!isDefined(key)) {
      return false;
    }

    return compareNumbers(previousData, data, 'increased', { key, actual, previous });
  };

  const removed = (key?: Key, value?: Value): boolean => {
    try {
      // skip if immediate parent isn't an array or object
      if (!isCompatible(previousData, data, { key })) {
        return false;
      }

      const [left, right] = getDataValues(previousData, data, { filter: true, key });

      if (value) {
        // skip if key exists in the narrowed data
        if (key && left[key] && right[key]) {
          return false;
        }

        return compareValues(right, left, value);
      }

      if (key && key in left && !(key in right)) {
        return true;
      }

      const [leftIterable, rightIterable] = getIterables(data, previousData, {
        key,
      });

      return leftIterable.length < rightIterable.length;
    } catch (error) {
      return false;
    }
  };

  return { added, changed, changedFrom, changedTo, decreased, emptied, filled, increased, removed };
}
