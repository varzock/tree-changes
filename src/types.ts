export type AcceptedTypes = string | boolean | number | PlainObject;
export type Comparator = Array<string | any[]>;
export type Data = PlainObject | AcceptedTypes[];
export type Key = string | number;
export type Value = AcceptedTypes | AcceptedTypes[];

export interface PlainObject {
  [key: string]: any;
}

export interface Params {
  actual?: Value;
  includeStrings?: boolean;
  filter?: boolean;
  key: Key;
  previous?: Value;
}

export type TreeChangesFn = (previousData: Data, data: Data) => TreeChanges;

export interface TreeChanges {
  added: (key?: Key, value?: Value) => boolean;
  changed: (key?: Key, actual?: Value, previous?: Value) => boolean;
  changedFrom: (key: Key, previous: Value, actual?: Value) => boolean;
  changedTo: (key: Key, actual: Value) => boolean;
  decreased: (key: Key, actual?: Value, previous?: Value) => boolean;
  emptied: (key?: Key) => boolean;
  filled: (key?: Key) => boolean;
  increased: (key: Key, actual?: Value, previous?: Value) => boolean;
  removed: (key?: Key, value?: Value) => boolean;
}
