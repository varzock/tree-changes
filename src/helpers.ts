import * as equal from 'fast-deep-equal';
import { Data, Key, Params, PlainObject, Value } from './types';

export function canHaveLength(...args: any): boolean {
  return args.every((d: any) => {
    return isString(d) || isArray(d) || isPlainObject(d);
  });
}

export function checkEntries(input: any) {
  return ([k, v]: [string, any]) => {
    if (equal(input, v)) {
      return true;
    }

    if (isArray(input)) {
      return input.some(d => equal(d, v) || (isArray(v) && v.indexOf(d) >= 0));
    }

    if (isPlainObject(input)) {
      return !!input[k] && equal(input[k], v);
    }

    return equal(input, v);
  };
}

export function checkItems(input: any) {
  return (value: any) => {
    if (isArray(input)) {
      return input.some(d => equal(d, value) || (isArray(value) && value.indexOf(d) >= 0));
    }

    return equal(input, value);
  };
}

export function compareNumbers(
  previousData: Data,
  data: Data,
  type: string,
  { actual, key, previous }: Params,
): boolean {
  const left = nested(previousData, key);
  const right = nested(data, key);
  const hasActual = typeof actual !== 'undefined';
  const hasPrevious = typeof previous !== 'undefined';

  let changed =
    [left, right].every(isNumber) && (type === 'increased' ? left < right : left > right);

  if (hasActual) {
    changed = changed && right === actual;
  }

  if (hasPrevious) {
    changed = changed && left === previous;
  }

  return changed;
}

export function compareValues(left: Data, right: Data, value: Value) {
  if (!isSameType(left, right)) {
    return false;
  }

  if (!isDefined(left) && right) {
    return true;
  }

  if (isArray(left) && isArray(right)) {
    return !left.some(checkItems(value)) && right.some(checkItems(value));
  }

  /* istanbul ignore else */
  if (isPlainObject(left) && isPlainObject(right)) {
    return (
      !Object.entries(left).some(checkEntries(value)) &&
      Object.entries(right).some(checkEntries(value))
    );
  }

  return right === value;
}

export function getIterables(
  previousData: Data,
  data: Data,
  { filter = true, includeStrings, key }: Partial<Params> = {},
) {
  let [left, right] = getDataValues(previousData, data, { filter, key });

  if (!isSameType(left, right)) {
    throw new TypeError('Inputs have different types');
  }

  if (!canHaveLength(left, right)) {
    throw new TypeError("Inputs dont't have length");
  }

  if (!includeStrings && [left, right].every(isString)) {
    throw new TypeError('Strings are excluded');
  }

  if ([left, right].every(isPlainObject)) {
    left = Object.keys(left);
    right = Object.keys(right);
  }

  return [left, right];
}

export function getDataValues(prevData: any, data: any, options?: Partial<Params>) {
  const { filter, key } = options || {};

  if (key) {
    const leftWithKey = nested(prevData, key);
    const rightWithKey = nested(data, key);

    if (
      !filter ||
      (isPlainObject(leftWithKey) && isPlainObject(rightWithKey)) ||
      (isArray(leftWithKey) && isArray(rightWithKey))
    ) {
      return [leftWithKey, rightWithKey];
    }
  }

  return [prevData, data];
}

export function getKeys(key: Key) {
  let keys: Array<string | number> = [];

  if (isString(key)) {
    keys = key.split('.');
  } else {
    keys.push(key);
  }

  return {
    parentKey: keys.slice(-2, 1)[0],
    childKey: keys.slice(-1)[0],
  };
}

export function isArray(value: unknown): value is Array<typeof value> {
  return Array.isArray(value);
}

export function isCompatible(prevData: any, data: any, options?: Partial<Params>): boolean {
  const { filter = true, includeStrings, key } = options || {};
  let upperKey;

  if (key) {
    const { childKey, parentKey } = getKeys(key);
    upperKey = parentKey || childKey;
  }

  const [left, right] = getDataValues(prevData, data, { filter, key: upperKey });

  if (!isSameType(left, right)) {
    return false;
  }

  const checks = [isPlainObject(left) && isPlainObject(right), isArray(left) && isArray(right)];

  if (includeStrings) {
    checks.push(isString(left) && isString(right));
  }

  return checks.some(d => d === true);
}

export function isDefined(value: unknown): boolean {
  return typeof value !== 'undefined';
}

export function isNumber(value: unknown): value is number {
  return typeof value === 'number';
}

export function isPlainObject(value: unknown): value is PlainObject {
  if (!value) {
    return false;
  }
  const prototype = Object.getPrototypeOf(value);

  return (
    Object.prototype.toString.call(value).slice(8, -1) === 'Object' &&
    (prototype === null || prototype === Object.getPrototypeOf({}))
  );
}

export function isString(value: unknown): value is string {
  return typeof value === 'string';
}

export function isSameType(...args: any): boolean {
  return (
    args.every(isArray) || args.every(isNumber) || args.every(isPlainObject) || args.every(isString)
  );
}

export function nested(input: PlainObject | any[], property?: Key) {
  /* istanbul ignore else */
  if (isPlainObject(input) || isArray(input)) {
    /* istanbul ignore else */
    if (isString(property) && property !== '') {
      const split = property.split('.');
      // @ts-ignore
      return split.reduce((obj, prop) => obj && obj[prop], input);
    }

    if (isNumber(property) && isArray(input)) {
      return input[property];
    }

    return input;
  }

  return input;
}
