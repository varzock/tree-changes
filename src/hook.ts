import { useEffect, useMemo, useRef } from 'react';

import treeChanges from './base';

export function usePrevious<T>(value: T): T | undefined {
  const ref = useRef<T>();

  useEffect(() => {
    ref.current = value;
  }, [value]);

  return ref.current;
}

export default function useTreeChanges(value: any) {
  const previousValue = usePrevious(value);

  return useMemo(() => treeChanges(previousValue || value, value), [previousValue, value]);
}
