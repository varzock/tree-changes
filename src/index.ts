import treeChanges from './base';

export default treeChanges;
