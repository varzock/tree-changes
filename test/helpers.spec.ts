// @ts-ignore
import {
  canHaveLength,
  compareNumbers,
  compareValues,
  getIterables,
  isArray,
  isDefined,
  isNumber,
  isPlainObject,
  isSameType,
  isString,
  nested,
} from '../src/helpers';

import { PlainObject } from '../src/types';

const left: PlainObject = {
  data: [],
  messages: [
    { id: 1, messsage: 'hello' },
    { id: 2, messsage: 'hey' },
    { id: 3, messsage: 'sup?' },
  ],
  name: '',
  options: {},
  ratio: 0.8,
  retries: 0,
  status: 'idle',
  version: 1,
};

const right: PlainObject = {
  data: [{ a: 1 }],
  messages: [
    { id: 2, messsage: 'hey' },
    { id: 3, messsage: 'sup?' },
  ],
  name: 'John',
  options: {
    ui: {
      color: '#333',
      tags: ['simple', 'clean'],
    },
    updatedAt: 1234567890,
  },
  ratio: 0.4,
  retries: 1,
  status: 'done',
  version: '1.1',
};

describe('canHaveLength', () => {
  it('should check properly', () => {
    expect(canHaveLength('0', '')).toBe(true);
    expect(canHaveLength([0], [])).toBe(true);
    expect(canHaveLength({}, { a: 1 })).toBe(true);

    expect(canHaveLength(1, 0)).toBe(false);
    expect(canHaveLength(1, {})).toBe(false);
  });
});

describe('compareNumbers', () => {
  describe('increased', () => {
    it('without options', () => {
      expect(compareNumbers(left, right, 'increased', { key: 'retries' })).toBe(true);

      expect(compareNumbers(left, right, 'increased', { key: 'data' })).toBe(false);
      expect(compareNumbers(left, right, 'increased', { key: 'options' })).toBe(false);
      expect(compareNumbers(left, right, 'increased', { key: 'ratio' })).toBe(false);
      expect(compareNumbers(left, right, 'increased', { key: 'stats' })).toBe(false);
    });

    it('with actual', () => {
      expect(compareNumbers(left, right, 'increased', { key: 'retries', actual: 1 })).toBe(true);
      expect(compareNumbers(left, right, 'increased', { key: 'retries', actual: 0.9 })).toBe(false);
      expect(compareNumbers(left, right, 'increased', { key: 'retries', actual: 10 })).toBe(false);
    });

    it('with actual and previous', () => {
      expect(
        compareNumbers(left, right, 'increased', { key: 'retries', actual: 1, previous: 0 }),
      ).toBe(true);
      expect(
        compareNumbers(left, right, 'increased', { key: 'retries', actual: 1, previous: 0.1 }),
      ).toBe(false);
      expect(
        compareNumbers(left, right, 'increased', { key: 'retries', actual: 1, previous: 10 }),
      ).toBe(false);
    });
  });

  describe('decreased', () => {
    it('without options', () => {
      expect(compareNumbers(left, right, 'decreased', { key: 'ratio' })).toBe(true);

      expect(compareNumbers(left, right, 'decreased', { key: 'data' })).toBe(false);
      expect(compareNumbers(left, right, 'decreased', { key: 'options' })).toBe(false);
      expect(compareNumbers(left, right, 'decreased', { key: 'retries' })).toBe(false);
      expect(compareNumbers(left, right, 'decreased', { key: 'stats' })).toBe(false);
    });

    it('with actual', () => {
      expect(compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 0.4 })).toBe(true);
      expect(compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 0.9 })).toBe(false);
      expect(compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 10 })).toBe(false);
    });

    it('with actual and previous', () => {
      expect(
        compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 0.4, previous: 0.8 }),
      ).toBe(true);
      expect(
        compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 0.4, previous: 0.1 }),
      ).toBe(false);
      expect(
        compareNumbers(left, right, 'decreased', { key: 'ratio', actual: 0.4, previous: 10 }),
      ).toBe(false);
    });
  });
});

describe('compareValues', () => {
  describe('left to right', () => {
    it.each([
      // Numbers
      [0.4, 'ratio', true],
      [1, 'retries', true],
      ['1', 'retries', false],

      // Objects
      // [{ color: '#333' }, 'options.ui', true],
      [{ updatedAt: 1234567890 }, 'options', true],
      [{ id: 3, messsage: 'sup?' }, 'messages', false],

      // Arrays
      [[{ a: 1 }], 'data', true],
      [[{ a: 1 }], 'data', true],
      // [{ a: 1 }, 'data', false],

      // String
      ['idle', 'status', false],
      ['done', 'status', true],
      ['', 'name', false],
      ['John', 'name', true],
    ])('compareValues %p should be %s)', (value, key, expected) => {
      expect(compareValues(left[key], right[key], value)).toBe(expected);
    });
  });

  describe('right to left', () => {
    it.each([
      [0.8, 'ratio', true],
      [{ id: 1, messsage: 'hello' }, 'messages', true],
      [{ id: 2, messsage: 'hey' }, 'messages', false],
      [{ id: 3, messsage: 'sup?' }, 'messages', false],
    ])('compareValues with %s should be %s)', (value, key, expected) => {
      expect(compareValues(right[key], left[key], value)).toBe(expected);
    });
  });
});

describe('getComparables', () => {
  it('should get iterables', () => {
    expect(getIterables(left, right, { key: 'data' })).toEqual([[], [{ a: 1 }]]);
    expect(getIterables(left, right, { key: 'options' })).toEqual([[], ['ui', 'updatedAt']]);
    expect(getIterables(left, right, { key: 'status' })).toMatchSnapshot();
  });

  it('should throw for invalid types', () => {
    expect(() => getIterables(left.version, right.version, { key: 'version' })).toThrow(
      'Inputs have different types',
    );

    expect(() => getIterables(left.ratio, right.ratio)).toThrow("Inputs dont't have length");

    expect(() => getIterables(left.status, right.status)).toThrow('Strings are excluded');
  });
});

describe('isArray', () => {
  it('should detect arrays', () => {
    expect([[], []].every(isArray)).toBe(true);

    expect([[], {}].every(isArray)).toBe(false);
    expect([0, []].every(isArray)).toBe(false);
    expect(['', 1].every(isArray)).toBe(false);
  });
});

describe('isDefined', () => {
  it('should return properly', () => {
    expect(isDefined('')).toBe(true);
    expect(isDefined(false)).toBe(true);
    expect(isDefined(null)).toBe(true);
    expect(isDefined([])).toBe(true);

    expect(isDefined(undefined)).toBe(false);
    // @ts-ignore
    expect(isDefined()).toBe(false);
  });
});

describe('isNumber', () => {
  it('should detect numbers', () => {
    expect([0, 1].every(isNumber)).toBe(true);

    expect([[], {}].every(isNumber)).toBe(false);
    expect([0, []].every(isNumber)).toBe(false);
    expect(['', 1].every(isNumber)).toBe(false);
  });
});

describe('isPlainObj', () => {
  it('should detect plain objects', () => {
    expect([{}, {}].every(isPlainObject)).toBe(true);

    expect([[], {}].every(isPlainObject)).toBe(false);
    expect([0, []].every(isPlainObject)).toBe(false);
    expect(['', 1].every(isPlainObject)).toBe(false);
  });
});

describe('isSameType', () => {
  it('should compare types', () => {
    expect(isSameType('0', '1')).toBe(true);
    expect(isSameType([1], [])).toBe(true);
    expect(isSameType({}, { a: 1 })).toBe(true);
    expect(isSameType(0, 1)).toBe(true);

    expect(isSameType('0', [])).toBe(false);
    expect(isSameType({}, [])).toBe(false);
    expect(isSameType(1, {})).toBe(false);
    expect(isSameType('0', 1)).toBe(false);
  });
});

describe('isString', () => {
  it('should detect string', () => {
    expect(['', 'null'].every(isString)).toBe(true);

    expect([[], {}].every(isString)).toBe(false);
    expect([0, []].every(isString)).toBe(false);
    expect(['', 1].every(isString)).toBe(false);
  });
});

describe('nested', () => {
  it('should get the correct data', () => {
    expect(nested(right, 'status')).toBe('done');
    expect(nested(right, 'options.ui')).toEqual({ color: '#333', tags: ['simple', 'clean'] });
    expect(nested(right, 'options.ui.tags.0')).toBe('simple');
    expect(nested(right.options.ui.tags, 1)).toBe('clean');
    expect(nested(right, 'options.updatedAt')).toBe(1234567890);
    // @ts-ignore
    expect(nested(right.status, 1)).toBe('done');
  });
});
