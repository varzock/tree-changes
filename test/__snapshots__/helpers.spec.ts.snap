// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComparables should get iterables 1`] = `
Array [
  Array [
    "data",
    "messages",
    "name",
    "options",
    "ratio",
    "retries",
    "status",
    "version",
  ],
  Array [
    "data",
    "messages",
    "name",
    "options",
    "ratio",
    "retries",
    "status",
    "version",
  ],
]
`;
